Fix #12046: skip publishing nightly build when no new commits since last publish

griggt opened the pull request:
Attempting to publish an already published version using `sbt-sonatype`
fails with a "Failed to promote the repository" error.

There don't appear to be any configuration settings in `sbt-sontaype`
for avoiding this, so instead we ask for published versions from the
Maven Central metadata and skip the publish step if we find a match.

Fixes #12046

griggt commented:
I did some testing of this proposal [over here](https://github.com/griggt/test-is-already-published/actions).

smarter commented on project/scripts/is-version-published.sh:
```diff
@@ -0,0 +1,47 @@
+
+set -eu
+
+major=${ver%%-bin-*}
```
Once https://github.com/lampepfl/dotty/pull/11952 is merged, the binary version will be fixed to 3, so this won't work anymore, I suggest waiting until that PR is merged then just using _3.

smarter replied:
https://repo1.maven.org/maven2/org/scala-lang/scala3-compiler_3/ is up, so you should be able to change the script to cut the version number after the first `.` or hardcode 3.

griggt commented:
I opted to use the version number up to the first dot as the binary version, in the hope that it may be less brittle, and renamed the var so as to be more likely to be found by `grep -i`. That said, the workflow is structured such that publication is only skipped when the script is successful, so even if it doesn't get updated when it should, it won't suddenly halt publishing nightlies.

I also changed the script to just do a HEAD request for the POM file rather than retrieve the maven-metadata.xml and iterate over the versions for a match. The latter approach is ok when there are a small number of published versions for each binary version but now that the binary version is stable the version list will just keep growing...

smarter reviewed (approved):
:sparkles:
